crates/matrix-sdk/src/sliding_sync.rs

```diff
@@ -339,23 +334,102 @@ impl SlidingSyncRoom {
+                //     [D, E, F]
+                // and if the `timeline_updates` contains:
+                //     [A, B, C, D, E, F]
+                // the resulting `timeline_queue` must be:
+                //     [A, B, C, D, E, F]
+                //
+                // To do that, we find the longest suffix between `timeline_queue` and
+                // `timeline_updates`, in this case:
+                //     [D, E, F]
+                // Remove the suffix from `timeline_updates`, we get `[A, B, C]` that is
+                // prepended to `timeline_queue`.
+                //
+                // If the `timeline_queue` contains:
+                //     [A, B, C, D, E, F]
+                // and if the `timeline_updates` contains:
+                //     [D, E, F]
+                // the resulting `timeline_queue` must be:
+                //     [A, B, C, D, E, F]
+                //
+                // To do that, we continue with the longest suffix. In this case, it is:
+                //     [D, E, F]
+                // Remove the suffix from `timeline_updates`, we get `[]`. It's empty, we don't
+                // touch at `timeline_queue`.
+
+                {
+                    let timeline_queue_len = timeline_queue.len();
+                    let timeline_updates_len = timeline_updates.len();
+
+                    let position = match timeline_queue
+                        .iter()
+                        .rev()
+                        .zip(timeline_updates.iter().rev())
+                        .map(|(queue, update)| (queue.event_id(), update.event_id()))
+                        .position(|(queue, update)| queue != update)
+                    {
+                        // We have found a suffix that equals the size of `timeline_queue` or
+                        // `timeline_update`, typically:
+                        //     timeline_queue = [D, E, F]
+                        //     timeline_update = [A, B, C, D, E, F]
+                        // or
+                        //     timeline_queue = [A, B, C, D, E, F]
+                        //     timeline_update = [D, E, F]
+                        // in both case, `position` will return `None` because we are looking (from
+                        // the end) an item that is different.
+                        None => std::cmp::min(timeline_queue_len, timeline_updates_len),
+
+                        // We have found a suffix but it doesn't cover all `timeline_queue` or
+                        // `timeline_update`, typically:
+                        //     timeline_queue = [B, D, E, F]
+                        //     timeline_update = [A, B, C, D, E, F]
+                        // in this case, `position` will return `Some(3)` instead of `None`.
+                        // That's annoying because it means we have an invalid `timeline_queue` or
+                        // `timeline_update`.
+                        Some(position) => position,
```

Comment:
"We have found a suffix" is wrong, the position can be `0` in which case there is no common suffix.

Reply:
:+1: I've improved the documentation to detail both case (where `Some(n)` has `n = 0` and `n > 0`).